### PR TITLE
[AIR - Datasets] Ragged tensor extension type.

### LINKED ISSUE #18316: [Datasets] Support heterogeneous shapes in tensor columns.
Certain use cases, such as a table column containing images (2D arrays), requires that the tensors be able to have different shapes (e.g. 128x128 and 256x256) within that column. We should extend our existing Pandas/Arrow extension types/arrays, or add a new Pandas/Arrow extension types/arrays, to support heterogeneous tensor shapes within tensor columns.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Ragged tensors pop up in NLP ([textual sentences/documents of different lengths](https://towardsdatascience.com/using-tensorflow-ragged-tensors-2af07849a7bd), [N-grams](https://www.tensorflow.org/guide/ragged_tensor#example_use_case)), CV (images of different resolutions, [`ssd300_vgg16` detection outputs](https://pytorch.org/vision/main/models/generated/torchvision.models.detection.ssd300_vgg16.html)) and audio (differing durations) use cases. Our existing tensor extension doesn't support ragged tensors, failing when the we or the user attempts to cast such data to it.

This PR adds support for non-nested ragged tensors, i.e. where each each tensor element (such as an image or sentence) is non-ragged. This allows us to support imagery columns of differing shapes or even differing number of dimensions, object detection outputs containing a variable number of detections (with a bounding box per detection) per sample, text columns containing N-gram embeddings, etc.

In order to keep the PR as simple as possible while meeting the requirements of these use cases, this PR does _not_ support arbitrary nesting of ragged tensors. Any such ragged tensor extension with arbitrary nesting should most likely be done in upstream Arrow, in C++.

## TODOs

- [x] More test coverage geared towards use cases.
- [x] Update feature guides.
- [ ] Update AIR examples that contain/need ragged tensor support.
- [x] Look to further consolidate `ArrowTensorType` and `ArrowRaggedTensorType`, `ArrowTensorArray` and `ArrowRaggedTensorArray`.

## Related issue number

Closes https://github.com/ray-project/ray/issues/18316

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/pandas.py:724]: Object-dtyped columns occasionally arise that _can_ be consolidated into a contiguous, well-typed ndarray, in which case we should do so since it makes everything far more efficient downstream.

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/pandas.py:1382]: Only checks the first tensor element as a best-effort O(1) check.

File: python/ray/air/util/tensor_extensions/arrow.py
Comment (clarkzinzow) [inline python/ray/air/util/tensor_extensions/arrow.py:335]: It may be better to call this `ArrowVariableShapeTensorArray` or something to that effect, since a "ragged" tensor supports arbitrary raggedness, e.g. any dimension can be ragged, while this tensor extension has a stricter requirement that all tensor elements must themselves be non-ragged.

File: python/ray/air/util/tensor_extensions/arrow.py
Comment (jiaodong) [inline python/ray/air/util/tensor_extensions/arrow.py:449]: if we're using `all` here for pairwise element, which i think is already O(n) operation, why not further safe guard this to ensure our `_is_ragged_tensor` function also checks shape of all element rather than just optimistically O(1) by checking the first element ?

From previous segmentation model CUJ i found it's quite common in larger dataset with blocks / table concat that the first element looks fine but last / middle element actually have different shape / type.

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (jiaodong) [inline python/ray/air/util/tensor_extensions/pandas.py:716]: i think it's better for us to define a few air tensor tension specific exceptions to narrowly define cases we explicitly do not support, so upper level air components can catch, make fall back decisions or print more actionable message based on context if needed.

Currently I found the exception we throw here are very general and not particularly actionable / customized for each AIR application type

File: python/ray/data/tests/test_dataset.py
Comment (jiaodong) [inline python/ray/data/tests/test_dataset.py:488]: can we also have a test that provides multi-dimension ragged ndarray and ensure it throw right exception type + message as expected ?

**Review by jiaodong — COMMENTED** [review]
great to see work on this and knowing we have a few cards in hand to ravel them into 1-D continuous array :)

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:716]: I believe that it's idiomatic and/or required to raise a `TypeError` from the extension array constructor when construction fails, due to how the Pandas type extension machinery works. I know that is at least true for the extension type.

I'll double-check that this is the case for the extension array as well.

File: python/ray/air/util/tensor_extensions/arrow.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/arrow.py:449]: A few things here:
1. This O(n) pairwise check is only done once, on construction, and we're already doing other linear passes here. The `_is_ragged_tensor()` check currently happens whenever the dtype is requested, hence the hope to keep it as an O(1) check. We could always cache the check on construction since `TensorArray` should be immutable.
2. The best-effort `_is_ragged_tensor()` check is only checking if the ndarray is of object dtype and its elements are ndarrays; it's not checking to see if all elements are of different shape. Since object-dtyped tensors can sneak in, this check probably needs to be more rigorous in general, e.g. checking that (1) the ndarray is of object dtype, (2) the elements are all ndarrays, (3) the elements all have the same dtype.

File: python/ray/air/util/tensor_extensions/arrow.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/arrow.py:336]: I didn't call it that since this ragged tensor extension can be ragged in more than a single dimension; e.g., if you have a tensor containing 2 images with shapes (28, 28) and (32, 32), that's ragged in both the height and width dimensions.

File: python/ray/air/util/tensor_extensions/arrow.py
Comment (jiaodong) [reply python/ray/air/util/tensor_extensions/arrow.py:453]: Agreed on the context of 1) , for 2) looks like there's no O(1) way to know all elements dtype based on your description. What if we adopt caching the check as metadata to the block if the immutability contract holds ? So upon seeing TensorArray we always know if it's ragged or not in subsequent processing.

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:726]: Hmm actually, I'm not seeing where AIR would benefit from multiplexing on specific exception types or adding more context to specific exception types.
1. We aren't doing try-to-cast fallbacks since trying to construct a tensor column is expensive, so we're embracing cheap LBYL since EAFP would be costly.
2. I don't see spots where we'd want to recast these errors with higher-level AIR context, due to (1).
3. Actionable error messages would be most useful to users doing their own tensor column casting/construction, so we should directly improve the `TensorArray` exception messages for those users.

Could we punt on this until we want to add such AIR-level fallbacks or more actionable AIR-level messages?

Comment (jiaodong) [reply python/ray/air/util/tensor_extensions/pandas.py:726]: +1 punt on this since we're not doing fallbacks anymore. We can revisit this once we got to the general simple block / opaque python object if the need arises from user

Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:726]: Sounds good!

File: python/ray/air/_internal/tensorflow_utils.py
Comment (jiaodong) [inline python/ray/air/_internal/tensorflow_utils.py:75]: i think both `tf.stack` and `tf.ragged.stack` raises `ValueError`:

https://www.tensorflow.org/api_docs/python/tf/stack
https://www.tensorflow.org/api_docs/python/tf/ragged/stack

**Review by jiaodong — COMMENTED** [review]
thanks for iterating on this ! I left a few nits but I think there're a lot of dataset internal changes that @c21 and @jianoaix can provide a lot more insights. 

Only item on my end is to verify this with our existing workloads (object detection for example), but this can be follow-up PR at later stage after critical PRs are merged before branch cut.

File: python/ray/air/_internal/tensorflow_utils.py
Comment (clarkzinzow) [reply python/ray/air/_internal/tensorflow_utils.py:75]: Hmm I thought that I saw other exceptions leak out, but I'll try reducing the exception scope to `ValueError` and see if it suffices!

File: doc/source/data/doc_code/tensor.py
Comment (clarkzinzow) [reply doc/source/data/doc_code/tensor.py:519]: `iter_torch_batches()` is the current canonical Torch route, but `iter_tf_batches()` is currently broken and would result in a very verbose example, and we're planning on making `to_tf()` the canonical TensorFlow route, so I opted for it here.

**Comment by clarkzinzow** [comment]
>Only item on my end is to verify this with our existing workloads (object detection for example), but this can be follow-up PR at later stage after critical PRs are merged before branch cut.

I plan on porting those examples while the general review of this PR is ongoing!

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (c21) [inline python/ray/air/util/tensor_extensions/pandas.py:1389]: why this is a public API? are we expecting users to use it? Should it be a developer API or without API annotation?

**Review by c21 — APPROVED** [review]
Thanks @clarkzinzow for the great work!

My knowledge of tensor-related code are still shallow, but this LGTM with my understanding so far. @jianoaix would be good to take a look as well, thanks.

File: python/ray/air/_internal/tensorflow_utils.py
Comment (clarkzinzow) [reply python/ray/air/_internal/tensorflow_utils.py:81]: I was thinking that the attempted ragged tensor construction would be very uncommon and that the original exception would most likely be higher signal, but I think you're right that including the context from this failed construction would be very useful for users trying to utilize ragged tensors.

How about we structure the exception chain as `f"Normal tensor construction failed, so we tried to construct a ragged tensor, which also failed.\nNormal tensor construction error: {outer_e}\n\nRagged tensor construction error: {e}"` or something like that?

File: python/ray/air/_internal/torch_utils.py
Comment (clarkzinzow) [reply python/ray/air/_internal/torch_utils.py:76]: IIRC `torch.stack()` raises a `RuntimeError` when passing a tensor that's coercible to a nested (ragged) tensor; I can re-run and add a comment.

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:1389]: Ah good point, this was back from before `TensorArray` could support both non-variable shaped and variable shaped tensors, when the user would need to do their own check to cast their tensor to the right extension type. I'll get rid of this annotation and make this a private utility.

File: python/ray/air/_internal/torch_utils.py
Comment (c21) [reply python/ray/air/_internal/torch_utils.py:76]: Cool, it can be done as followup. Not urgent.

**Comment by c21** [comment]
Hi @jianoaix - any more comments for the PR?

**Review by jianoaix — COMMENTED** [review]
Mostly delegating the details of tensor type/array to previous review by Jiao and Cheng.

File: python/ray/air/_internal/tensorflow_utils.py
Comment (clarkzinzow) [reply python/ray/air/_internal/tensorflow_utils.py:75]: In any case, we want this to be a blanket errors catch-all anyways, where we want to:
1. try to convert, and if that fails ->
2. try to convert individual elements and stack, and if that fails ->
3. try to construct a ragged tensor, and if that fails ->
4. raise an error capturing our attempted conversions and their respective failures.

If (2) and (3) don't have a well-defined set of exception types for failures modes on data we expect to see, I think that a blanket `except Exception` catch-all is fine.

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:1390]: This is heavily tied to the internal variable-shaped tensor representation of the Pandas-side of the tensor extension type (which uses a NumPy ndarray for its internal representation), so I was planning on housing it next to the extension type since (1) it uses it internally to determine whether its variable-shaped or not, and (2) we need this utility and the `TensorDtype` and `TensorArray` representations to stay in sync.

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:1389]: Hmm this would still be needed by users if they're implementing their own Arrow UDFs + using possible-ragged tensor columns. I'll see if I can hide this.

File: python/ray/air/util/tensor_extensions/arrow.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/arrow.py:341]: Hmm I thought that it was clear that homogeneous-typed means homogeneous types, i.e. homogeneous dtypes? Would it be clearer if it was "homogeneous-dtyped"?

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (clarkzinzow) [reply python/ray/air/util/tensor_extensions/pandas.py:1390]: I opted to move it to a `tensor_extensions/utils.py` so both `ArrowTensorArray` and `TensorArray` can use it and we can hide it from the user.

**Comment by jiaodong** [comment]
Just to confirm regarding numpy narrow waist -- is there any internal arrow concatenation / chucked array for this new extension type that we need to be testing ? Similarly we should create a follow-up issue to add ragged numpy array test cases as part of tech debt pay off / internal documentation.

**Comment by clarkzinzow** [comment]
@jiaodong Great question! That should be taken care of by the generic extension type concatenation machinery here:
- https://github.com/ray-project/ray/blob/4befbbb7309785c394b4ff70eb06a0b33a0b3d96/python/ray/data/_internal/arrow_block.py#L212-L213
- https://github.com/ray-project/ray/blob/4befbbb7309785c394b4ff70eb06a0b33a0b3d96/python/ray/data/_internal/arrow_ops/transform_pyarrow.py#L19-L43

I'll double-check that this ragged tensor type is covered by tests for this.

File: python/ray/air/util/tensor_extensions/pandas.py
Comment (c21) [reply python/ray/air/util/tensor_extensions/pandas.py:1390]: >I opted to move it to a tensor_extensions/utils.py so both ArrowTensorArray and TensorArray can use it and we can hide it from the user.

Yeah, this sounds good to me.

File: python/ray/air/_internal/tensorflow_utils.py
Comment (c21) [reply python/ray/air/_internal/tensorflow_utils.py:75]: Ok, I think we can start with `except Exception`.

**Review by c21 — APPROVED** [review]

**Review by jianoaix — APPROVED** [review]

File: doc/source/data/dataset-tensor-support.rst
Comment (clarkzinzow) [reply doc/source/data/dataset-tensor-support.rst]: In the below example, a "multi-column tensor dataset" means all columns are tensor-typed. I was going to add a line about a "tensor dataset" concept, but I'm going to hold off on that since (1) that's a larger refactoring of this tensor support doc, and (2) I don't know if we're internally aligned on the messaging there yet.

**Comment by clarkzinzow** [comment]
Tests are passing, will add more tests + port examples in a follow-up PR. Merging!

**Pull request merged**
